colorbleed/plugins/maya/publish/validate_animation_out_set_related_node_ids.py

```diff
@@ -3,87 +3,127 @@
     optional = True
 
     ignore_types = ("constraint",)
 
     def process(self, instance):
         """Process all meshes"""
 
-        # Ensure all nodes have a cbId
+        # Ensure all nodes have a cbId and a related ID to the original shapes
+        # if a deformer has been created on the shape
         invalid = self.get_invalid(instance)
         if invalid:
             raise RuntimeError("Nodes found with non-related "
                                "asset IDs: {0}".format(invalid))
 
     @classmethod
-    def get_pointcache_nodes(cls, instance):
-
-        # Get out_SET
-        sets = cmds.ls(instance, type='objectSet')
-        pointcache_sets = [x for x in sets if x == 'out_SET']
-
-        nodes = list()
-        for s in pointcache_sets:
-            members = cmds.sets(s, query=True)
-            members = cmds.ls(members, long=True)  # ensure long names
-            descendents = cmds.listRelatives(members,
-                                             allDescendents=True,
-                                             fullPath=True) or []
-            descendents = cmds.ls(descendents,
-                                  noIntermediate=True,
-                                  long=True)
-            hierarchy = members + descendents
-            nodes.extend(hierarchy)
-
-        # ignore certain node types (e.g. constraints)
-        ignore = cmds.ls(nodes, type=cls.ignore_types, long=True)
-        if ignore:
-            ignore = set(ignore)
-            nodes = [node for node in nodes if node not in ignore]
-
-        return nodes
-
-    @classmethod
     def get_invalid(cls, instance):
-        invalid_items = []
+        """Get all nodes which do not match the criteria"""
+        invalid = []
 
         # get asset id
-        nodes = cls.get_pointcache_nodes(instance)
+        nodes = instance.data.get("out_hierarchy", instance[:])
         for node in nodes:
+            node_type = cmds.nodeType(node)
             node_id = lib.get_id(node)
+
+            if node_type == "mesh" and not node_id:
+                invalid.append(node)
+                continue
+
             if not node_id:
-                invalid_items.append(node)
+                continue
+
+            root_id = cls.get_history_root_id(node=node)
+            if root_id is not None:
+                asset_id = cls.to_item(node_id)
+                if root_id != asset_id:
+                    invalid.append(node)
+
+        return invalid
+
+    @classmethod
+    def get_history_root_id(cls, node):
+        """
+        Get the original node ID when a node has been deformed
+
+        Args:
+            node (str): node to retrieve the
+
+        Returns:
+            str: the asset ID as found in the database
+
+        """
+
+        node = cmds.ls(node, long=True)[0]
+
+        # We only check when the node is *not* referenced
+        if cmds.referenceQuery(node, isNodeReferenced=True):
```

Comment:
Shouldn't this logic of skipping to check this specific node be in `get_invalid` just like you're skipping others there too?

Reply:
So move this check to `get_invalid` instead of `get_history_root_id`?

Reply:
To me that makes more sense, yes. And the `get_history_root_id` could even be a generic function, just like `get_parent`. It doesn't require any information of the class or the instance - in a way it's simple and elegant.